homework/grayscale-image/compression.cpp

```diff
@@ -0,0 +1,79 @@
+#include "compression.hpp"
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+
+CompressedBitmap compressGrayscale(const Bitmap& bitmap) {
+    CompressedBitmap compressed(width * height, std::make_pair(0, 0));
+    uint8_t counter = 0;
+    uint8_t counterLine = 0;
+    auto iter = begin(bitmap)->begin();
+    std::transform(iter, 
+                   (end(bitmap) - 1)->end(), 
+                   begin(compressed), 
+                   [&](auto pixel) { 
+                       if (counterLine == height){
+                           counterLine = 0;
+                           counter = 0;
+                       }
+                       if (pixel != *std::prev(iter)) {
+                           counter = 1;
+                       } else {
+                           counter++;
+                       }                       
+                       counterLine++;
+                       iter++;
+                       return std::make_pair(pixel, counter); 
+                   });
+    auto it = begin(compressed);
+    compressed.erase(std::remove_if(begin(compressed), 
+                                    end(compressed), 
+                                    [&](auto ele){ 
+                                        it++; 
+                                        return ele.first == (*it).first && ele.second == (*it).second - 1; }), 
+                     end(compressed));
+    return compressed;
+}
+
+Bitmap decompressGrayscale(const CompressedBitmap& compressed) {
+    uint8_t pixelsAmount = 0;
+    Bitmap bitmap{};
+    auto it = begin(compressed);
+    std::transform(begin(bitmap)->begin(), 
+                   (end(bitmap) - 1)->end(), 
+                   begin(bitmap)->begin(), 
+                   [&](auto pixel){ 
+                       if (it->second - pixelsAmount == 0) {
+                           pixelsAmount = 0;
+                           it++;
+                       }
+                       pixelsAmount++;
+                       return it->first; });
+    return bitmap;
```

Comment:
std::fil_n is enough. The array is continuous in memory, so std::fill_n will automatically jump to the next line.